pool.go

```diff
@@ -24,28 +24,85 @@ import (
 )
 
 type modulePool[T interfaces.Signature] struct {
-	pool sync.Pool
-	new  func() (*module[T], error)
+	pool    sync.Pool
+	maxSize uint32
+	new     func(*module[T]) (*module[T], error)
+	close   func(*module[T])
+	ch      chan *module[T] // for buffered channel impl
 }
 
-func newModulePool[T interfaces.Signature](ctx context.Context, template *template[T]) *modulePool[T] {
-	return &modulePool[T]{
-		new: func() (*module[T], error) {
-			return newModule[T](ctx, template)
-		},
+func newModulePool[T interfaces.Signature](ctx context.Context, template *template[T], maxSize uint32) *modulePool[T] {
+	if maxSize == 0 {
+		// If size = 0, use a standard sync.Pool + golang's runtime finalizers
+		// to make sure the close function gets called eventually.
+		return &modulePool[T]{
+			maxSize: maxSize,
+			new: func(m *module[T]) (*module[T], error) {
+				m.SetFinalizer()
+				return newModule[T](ctx, template)
+			},
```

Comment:
As far as I can tell, calls to modulePool.new actually pass null pointers, so it seems like this should be panicing. This approach seems closer.
```suggestion
			new: func() (*module[T], error) {
			        m := newModule[T](context.Background(), template)
				m.SetFinalizer()
				return m, nil
			},
```